### PR TITLE
feat: update GPU metrics in NodeMetric CRD

### LINKED ISSUE #370: [proposal] koordlet updates device resource allocation status in NodeMetric
**What is your proposal**:
Reporter in koordlet needs to update resources information and healthy status per device including GPU, RDMA and FPGA, etc on NodeMetric CRD based on metrics collected by metrics-advisor.

**Why is this needed**:
koord-scheduler needs info in this CRD to make scheduling decisions.

**Is there a suggested solution, if so, please add it**:

### PR DESCRIPTION
**Pull request by LambdaHJ**
Signed-off-by: 黄金 <huangjin_henan@outlook.com>

### Ⅰ. Describe what this PR does
add gpu metrics to crd.

### Ⅱ. Does this pull request fix one issue?
fixes #370

### V. Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `make test`

### CODE REVIEW

File: pkg/koordlet/metricsadvisor/collector_gpu.go [hunk 1]
```diff
@@ -197,17 +205,29 @@ func (g *gpuDeviceManager) getTotalGPUUsageOfPIDs(pids []uint64) []metriccache.G
...
 	}
 	return rtn
 }
 
 func (g *gpuDeviceManager) getPodGPUUsage(podParentDir string, cs []corev1.ContainerStatus) ([]metriccache.GPUMetric, error) {
+	runingContainer := make([]corev1.ContainerStatus, 0)
+	for i := range cs {
+		if cs[i].State.Running == nil {
+			klog.V(6).Infof("non-running container %s", cs[i].ContainerID)
+			continue
```
Comment (jasonliu747) [inline pkg/koordlet/metricsadvisor/collector_gpu.go:224]: according to this [convention](https://github.com/koordinator-sh/koordinator/blob/main/docs/contributors/logging.md), I think it should be V(5). WDYT?

File: pkg/koordlet/metricsadvisor/collector_gpu.go [hunk 2]
```diff
@@ -216,6 +236,10 @@ func (g *gpuDeviceManager) getPodGPUUsage(podParentDir string, cs []corev1.Conta
 }
 
 func (g *gpuDeviceManager) getContainerGPUUsage(podParentDir string, c *corev1.ContainerStatus) ([]metriccache.GPUMetric, error) {
+	if c.State.Running == nil {
+		klog.V(6).Infof("non-running container %s", c.ContainerID)
```
Comment (jasonliu747) [inline pkg/koordlet/metricsadvisor/collector_gpu.go:240]: same above, V(5).

File: pkg/koordlet/reporter/reporter.go [hunk 3]
```diff
@@ -354,19 +357,57 @@ func (su *statusUpdater) updateStatus(nodeMetric *slov1alpha1.NodeMetric, newSta
...
+			memoryRatioRaw := 100 * float64(gpu.MemoryUsed.Value()) / float64(gpu.TotalMemory.Value())
+			gpuInfo := schedulingv1alpha1.DeviceInfo{
+				UUID:  gpu.DeviceUUID,
+				Minor: gpu.Minor,
+				Type:  schedulingv1alpha1.GPU,
+				// TODO: how to check the health status of GPU
+				Resources: map[corev1.ResourceName]resource.Quantity{
+					apiext.GPUCore:        *resource.NewQuantity(int64(gpu.SMUtil), resource.DecimalExponent),
+					apiext.GPUMemory:      gpu.MemoryUsed,
+					apiext.GPUMemoryRatio: *resource.NewQuantity(int64(memoryRatioRaw), resource.DecimalExponent),
```
Comment (jasonliu747) [inline pkg/koordlet/reporter/reporter.go:372]: any reason for using `DecimalExponent`?

**Review by jasonliu747 — COMMENTED** [review]
nice work, only few comments.

File: pkg/koordlet/metricsadvisor/collector_gpu.go [hunk 4]
```diff
@@ -197,17 +205,29 @@ func (g *gpuDeviceManager) getTotalGPUUsageOfPIDs(pids []uint64) []metriccache.G
...
 		}
 	}
 	return rtn
 }
 
 func (g *gpuDeviceManager) getPodGPUUsage(podParentDir string, cs []corev1.ContainerStatus) ([]metriccache.GPUMetric, error) {
+	runingContainer := make([]corev1.ContainerStatus, 0)
+	for i := range cs {
+		if cs[i].State.Running == nil {
+			klog.V(6).Infof("non-running container %s", cs[i].ContainerID)
```
Comment (jasonliu747) [inline pkg/koordlet/metricsadvisor/collector_gpu.go:223]:
```
for _, c := range cs {
    if c.Stat.Running == nil {....}
}
```

File: pkg/koordlet/metricsadvisor/collector_gpu.go [hunk 5]
```diff
@@ -197,17 +205,29 @@ func (g *gpuDeviceManager) getTotalGPUUsageOfPIDs(pids []uint64) []metriccache.G
...
+				MemoryUsed:  *resource.NewQuantity(int64(value.MemoryUsed), resource.BinarySI),
+				TotalMemory: *resource.NewQuantity(int64(g.devices[i].TotalMemory), resource.BinarySI),
 			})
 		}
 	}
 	return rtn
 }
 
 func (g *gpuDeviceManager) getPodGPUUsage(podParentDir string, cs []corev1.ContainerStatus) ([]metriccache.GPUMetric, error) {
+	runingContainer := make([]corev1.ContainerStatus, 0)
```
Comment (jasonliu747) [inline pkg/koordlet/metricsadvisor/collector_gpu.go:220]: `runingContainer` - > `runningContainer`

File: pkg/koordlet/reporter/reporter.go (hunk 3 above)
Comment (LambdaHJ) [reply pkg/koordlet/reporter/reporter.go:372]: Use DecimalSI.

**Comment by jasonliu747** [comment]
这一部分的 UT 辛苦帮忙顺手补上～
image

**Review by jasonliu747 — APPROVED** [review]
/lgtm

**Pull request merged**